### ISSUE TITLE
Definining the max-len rule with ignorePattern containing a colon inline in a file results in a JSON parsing error

### ISSUE THREAD

**Issue by captainbrosset** [eslint/eslint#6224]
**What version of ESLint are you using?**
2.9.0

**What parser (default, Babel-ESLint, etc.) are you using?**
default

**Please show your full configuration:**
In .eslintrc I have configured the max-len rule to be:
`"max-len": [2, 80, 2, {"ignoreUrls": true, "ignorePattern": "data:image\/|\\s*require\\s*\\(|^\\s*loader\\.lazy|-\\*-"}],`

**What did you do? Please include the actual source code causing the issue.**
I wanted to experiment with 100 long lines in just one file, and didn't want to change the global rule config in .eslintrc. So I tried added the following comment at the top of the file:

`/* eslint max-len: [2, 100, 2, {ignoreUrls: true, "ignorePattern": "data:image\/|\\s*require\\s*\\(|^\\s*loader\\.lazy|-\\*-"}] */`

**What did you expect to happen?**
I wanted the 100 config to be taken into consideration and the ignorePattern to be applied too.

**What actually happened? Please include the actual, raw output from ESLint.**
The following error message appears:

`Failed to parse JSON from ' "max-len": [2, 100, 2, {"ignoreUrls": true, "ignorePattern": "data":image\/|\\s*require\\s*\\(|^\\s*loader\\.lazy|-\\*-"}]': Unexpected token d`

It looks like the code in eslint that reads rule configuration inline in files tries to convert it to valid JSON and in doing so replaces `:` with `":` which, in my case, actually breaks the JSON because of the regex I use in ignorePattern which includes a `:` character.

**Comment by mysticatea** [comment]
Thank you for this issue.

I confirmed it.
I guess the cause is this line: https://github.com/eslint/eslint/blob/777941e7ba6c1424572380edf5fd1b19b7c0e07c/lib/eslint.js#L83

I'm not sure how to fix this...

**Comment by captainbrosset** [comment]
Indeed, it does look like this line is responsible for this bug.
So, any property value that has a colon in it will hit the bug:

`{a:"foo:bar"}` will produce `{"a":""foo":bar"}`

One option would be to not try and modify the string at all and only accept valid JSON, just like in .eslintrc. But I'm guessing that would cause compatibility issues.

**Comment by mysticatea** [comment]
Another option is to make a fallback which parses it as valid JSON.

```js
    var convertedString = string.replace(/([a-zA-Z0-9\-\/]+):/g, "\"$1\":").replace(/(\]|[0-9])\s+(?=")/, "$1,");
    try {
        items = JSON.parse("{" + convertedString + "}");
    } catch (ex) {
        try {
            items = JSON.parse(string); // maybe also `require("./option").parse(["", "", "--rule", string]).rule` is OK.
        } catch (ex2) {
```

@eslint/eslint-team does anybody have any insight here?

**Comment by nzakas** [comment]
I believe @ilyavolodin wrote this bit initially.

**Comment by ilyavolodin** [comment]
I might've. I really don't remember now:-) I think the reason why that regular expression exists is that comments are usually not in a valid JSON form, so we need to quote key-names. What we might want to do is modify RegExp, so it wouldn't ignore colons that are already inside quotes. Or @mysticatea solution would work (although would be slower in the case of edge case like the one described above).

**Comment by platinumazure** [comment]
@nzakas @ilyavolodin Is it worth using levn to parse the strings instead? If nothing else, it means that (discounting shell escapes) the CLI syntax for toggling rules one-off (i.e., the `--rule` option) and the syntax for inline configuration would be the same. That could be a breaking change, though.

**Comment by mysticatea** [comment]
Actually, I tried levn, but I could not replace it by levn without breaking changes -- currently we can omit commas between rules, but levn cannot omit it. `/*eslint no-alert:2 no-console:2*/`

I think we can:

1. try parsing by levn.
2. do the current parsing as a fallback.
3. (remove the fallback in v3)

**Comment by platinumazure** [comment]
Well, before we commit to levn we should make sure that the OP's use case would be helped by that maneuver. (I apologize, I haven't had time to test it.)

**Comment by nzakas** [comment]
Let's shoot for a non-breaking change first. This is simple enough that a really basic tokenizer could handle it. We can try using Espree's tokenizer and write a small parser on top of it.

**Comment by michaelficarra** [comment]
Can't you parse it using `js-yaml`, which is already a dependency?
